java/ql/src/utils/modelgenerator/internal/CaptureTypeBasedSummaryModels.qll

```diff
@@ -0,0 +1,333 @@
+      return.getName().matches("Iterator%") and
+      et = return.(ParameterizedType).getTypeArgument(0) and
+      t.getTypeParameter(position) = et and
+      instantiates(this, t, position, elementType)
+    )
+  }
+
+  /**
+   * Returns the iterator element type of `this`.
+   */
+  Type getElementType() { result = elementType }
+}
+
+/**
+ * Holds if type `bound` is an upper bound for type `t` or equal to `t`.
+ */
+private predicate isEffectivelyUpperBound(Type t, Type bound) {
+  t = bound or
+  t.(Wildcard).getUpperBound().getType() = bound
+}
+
+/**
+ * Holds if type `bound` is a lower bound for type `t` or equal to `t`.
+ */
+private predicate isEffectivelyLowerBound(Type t, Type bound) {
+  t = bound or
+  t.(Wildcard).getLowerBound().getType() = bound
+}
+
+/**
+ * Holds if `t` is a container like type of `tv` (eg. `List<T>`).
+ * Note that collections are covariant in their element type.
+ */
+private predicate genericContainerType(RefType t, TypeVariable tv) {
+  exists(Type et |
+    et =
+      [
+        t.(ContainerType).getElementType(), t.(IterableType).getElementType(),
+        t.(Array).getElementType()
+      ]
+  |
+    isEffectivelyUpperBound(et, tv)
+  )
+}
+
+/**
+ * Holds if `tv` is a type variable of the immediate type declaring `callable`.
+ */
+private predicate classTypeParameter(Callable callable, TypeVariable tv) {
+  callable.getDeclaringType().(GenericType).getATypeParameter() = tv
+}
+
+/**
+ * Holds if `tv` is type variable of `callable` or the type declaring `callable`.
+ */
+private predicate localTypeParameter(Callable callable, TypeVariable tv) {
+  classTypeParameter(callable, tv) or
+  callable.(GenericCallable).getATypeParameter() = tv
+}
+
+/**
+ * Gets the access path postfix for `t`.
+ */
+private string getAccessPath(Type t) {
+  if
+    t instanceof Array and
+    not Specific::isPrimitiveTypeUsedForBulkData(t.(Array).getElementType())
+  then result = ".ArrayElement"
+  else
+    if t instanceof ContainerType or t instanceof IterableType
+    then result = ".Element"
+    else result = ""
+}
+
+/**
+ * Gets the access path for parameter `p`.
+ */
+private string parameterAccess(Parameter p) {
+  result = "Argument[" + p.getPosition() + "]" + getAccessPath(p.getType())
+}
+
+/**
+ * Holds if `callable` has a type parameter `tv` or container parameterized over type `tv`.
+ */
+private predicate parameter(Callable callable, string input, TypeVariable tv) {
+  exists(Parameter p, Type pt |
+    input = parameterAccess(p) and
+    p = callable.getAParameter() and
+    pt = p.getType() and
+    (
+      // Parameter of type tv
+      isEffectivelyUpperBound(pt, tv)
+      or
+      // Parameter is a container of type tv
+      genericContainerType(pt, tv)
+    )
+  )
+}
+
+/**
+ * Gets the string representation of a synthetic field corresponding to `tv`.
+ */
+private string getSyntheticField(TypeVariable tv) {
+  result = ".SyntheticField[ArgType" + tv.getIndex() + "]"
```

Comment:
This will need some refinement at some point, but it's probably good enough as a starting point.

Reply:
Yes, we do not dig into the inheritance hierarchy at the moment, but that needs to be improved.